src/python/pants/backend/tools/semgrep/rules.py

```diff
@@ -0,0 +1,230 @@
+from pants.engine.process import FallibleProcessResult, ProcessCacheScope
+from pants.engine.rules import Get, MultiGet, Rule, collect_rules, rule
+from pants.engine.target import AllTargets, Target, Targets
+from pants.engine.unions import UnionRule
+from pants.option.global_options import GlobalOptions
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize, softwrap
+
+from .subsystem import SemgrepFieldSet, SemgrepSubsystem
+from .target_types import SemgrepRuleSourceField
+
+logger = logging.getLogger(__name__)
+
+
+class SemgrepLintRequest(LintTargetsRequest):
+    field_set_type = SemgrepFieldSet
+    tool_subsystem = SemgrepSubsystem
+
+
+@dataclass(frozen=True)
+class PartitionMetadata:
+    config_files: frozenset[SemgrepRuleSourceField]
+    ignore_files: Snapshot
+
+    @property
+    def description(self) -> str:
+        return ", ".join(sorted(field.file_path for field in self.config_files))
+
+
+_IGNORE_FILE_NAME = ".semgrepignore"
+
+
+def warn_about_ignore_files_if_required(ignore_files: Snapshot, semgrep: SemgrepSubsystem) -> None:
+    non_root_files = sorted(name for name in ignore_files.files if name != _IGNORE_FILE_NAME)
+    if non_root_files and not semgrep.acknowledge_nested_semgrepignore_files_are_not_used:
+        # https://github.com/returntocorp/semgrep/issues/5669
+        logger.warning(
+            softwrap(
+                f"""
+                Semgrep does not obey {_IGNORE_FILE_NAME} outside the working directory, which is
+                the build root when run by pants. These files may not have the desired effect:
+                {', '.join(non_root_files)}
+
+                Set `acknowledge_nested_semgrepignore_files_are_not_used = true` in the `[semgrep]`
+                section of pants.toml to silence this warning.
+                """
+            )
+        )
+
+
+@dataclass
+class SemgrepIgnoreFiles:
+    snapshot: Snapshot
+
+
+@dataclass
+class AllSemgrepConfigs:
+    targets: dict[str, list[Target]]
+
+    def ancestor_targets(self, address: Address) -> Iterable[Target]:
+        # TODO: introspect the semgrep rules and determine which (if any) apply to the files, e.g. a #
+        # Python file shouldn't depend on a .semgrep.yml that doesn't have any 'python' or 'generic' #
+        # rules, and similarly if there's path inclusions/exclusions.
+        # TODO: this would be better as actual dependency inference (e.g. allows inspection, manual
+        # addition/exclusion), but that can only infer 'full' dependencies and it is wrong (e.g. JVM
+        # things break) for real code files to depend on this sort of non-code linter config; requires
+        # dependency scopes or similar (https://github.com/pantsbuild/pants/issues/12794)
+        spec = address.spec_path
+
+        while True:
+            yield from self.targets.get(spec, [])
+
+            if not spec:
+                break
+
+            spec = os.path.dirname(spec)
```

Comment:
use `pathlib` utilities? :smile:

Reply:
Ah, two unfortunate subtleties: `list(Path("abc")) == [Path(".")]`, which doesn't include `abc`, and `"."` needs to be `""` for convenient look-up. It's probably still better than this `while` loop though.

Reply:
# Instead of `chain` I usually see `[spec, *path.parents()`